dclab/rtdc_dataset/plugins/plugin_feature.py

```diff
@@ -1,27 +1,65 @@
+"""
+.. versionadded:: 0.34.0
+"""
 
-# Create PlugInFeature class which is child of AncillaryFeature
-# it should have methods that can be overridden by a script
-# The script should create a child class of PlugInFeature
-# The methods defined in the script class will override the
-# PlugInFeature class.
-
-import abc
+import pathlib
+import importlib
+import sys
 
 from ..ancillaries import AncillaryFeature
 
 
-class PlugInFeature(AncillaryFeature, abc.ABC):
-    def __init__(self, *args, **kwargs):
-        super(PlugInFeature, self).__init__(*args, **kwargs)
+_registered_plugin_instances = []
+
+
+def create_new_plugin_feature(plugin_path):
+    """find an instanciate a PlugInFeature from a user-defined script"""
+    info = find_plugin_feature_script(plugin_path)
+    plugin_list = []
+    for ii in range(len(info["feature names"])):
+        ref_info = {
+            "feature_name": info["feature names"][ii],
+            "method": info["method"],
+            "req_config": info["config required"][ii],
+            "req_features": info["features required"][ii],
+            "priority": info["priority"],
+        }
+        plugin_list.append(PlugInFeature(**ref_info))
+        # add feature label etc
+    return plugin_list
+
+
+def find_plugin_feature_script(plugin_path):
+    # find script, return info dict
+    path = pathlib.Path(plugin_path)
+    # insert the plugin directory to sys.path so we can import it
+    sys.path.insert(-1, str(path.parent))
```

Comment:
> # I am not sure whether we should support something like this:
> dclab.load_plugin_features_from_folder("/path/to/folder_containing_plugin_features/", recursive=True)
> # The problem with this is that this potentially leads to many errors if one plugin does not work and the user is overwhelmed.

This would be really nice. I feel like it wouldn't be difficult for `glob` to handle. I will see how many errors it throws up...


> I think dclab should not have any default directories for plugins. This should be handled by higher-level software like Shape-Out that can then also depend on e.g. `appdirs` or similar.

I agree, I was more hinting that we could tell users to have one place for all their plugins. Autoload with ShapeOut2 would be awesome.

I forgot to mention that I will be adding many more tests. Do you think loading the `examples/plugin.py` file would be okay to do for testing, or will I include a plugin in the test folder?

Reply:
It seems to not work without modifying the `sys.path`. Also, as far as I understand `finally` is always excecuted, so we should not `pop()` in the `finally` clause.